### ISSUE TITLE
Add the ability to evaluate the uncertainty at any point?

### ISSUE THREAD

**Issue by alconley** [geo-ant/varpro#29]
Hello, love this crate but it is missing a function to evaluate the uncertainty at any point (essentially a function like [lmfit's](https://lmfit.github.io/lmfit-py/model.html#lmfit.model.ModelResult.eval_uncertainty) eval_uncertainty  function). The function is based off of this [example](https://www.astro.rug.nl/software/kapteyn/kmpfittutorial.html#confidence-and-prediction-intervals). The python code of the function is located at line 1627 in model.py or [here](https://github.com/lmfit/lmfit-py/blob/master/lmfit/model.py) if you want to see how they do it!

The example shows you how to get the confidence and prediction intervals from the degrees of freedom, reduced chi squared, covariance matrix and best-fit parameters. Lmfit calculated the derivatives "by hand" whereas you could make the user give the derivatives to simplify.  

Adding something of this nature would be a great addition to varpro imo!

**Comment by geo-ant** [comment]
Hi, I'm happy you like the crate :) 
I have to check out the code you linked. Can you elaborate a little how the requested statistical info is different from [this here](https://docs.rs/varpro/latest/varpro/solvers/levmar/struct.LevMarSolver.html#method.fit_with_statistics) and [this](https://docs.rs/varpro/latest/varpro/statistics/struct.FitStatistics.html)?

**Comment by alconley** [comment]
So the other statistical information can be constructed with what is in FitStatistics. However maybe you could add them in so they are more easily obtained. The first would be 
1) Degrees of Freedom = N_observations - N_linear_paramaters - N_nonlinear_parameters
2) Chi-squared = sum of the squared weighted residuals
3) reduced chi-squared = chi-squared/degree of freedom

The main additional stat info that would be nice is a confidence/prediction interval at some given. Basically, if I have a double exponential fit, I want to prediction want my uncertainty would be at any position given some significance level. Likewise for a gaussian function and other fits. I will now show a example of how I have done this for a double exponential fit.

In the example I am creating a 1 sigma (68.3%) confidence band to predict the uncertainty for the fit from 0 to 7000 for a function of the form f(x)=ae^(-x/b)+ce^(-x/d). The first step is to get the studentT value for the degree of freedom given the significance level. I physically had to type the partial derivatives whereas lmfit calculates them (maybe its better that way?). I hope the code makes sense but I think it would be nice if you had a function in the [FitStatistics](https://docs.rs/varpro/latest/varpro/statistics/struct.FitStatistics.html#) that takes an x value (f64), defaults the significance level to 1 sigma but could be changed, and then returns a x and y value. However, I am not sure how you can recover the initial functions used to create the fit in the FitStatistics. 
```
use nalgebra::DVector;
use std::f64::consts::SQRT_2;
use varpro::model::builder::SeparableModelBuilder;
use varpro::solvers::levmar::{LevMarProblemBuilder, LevMarSolver};

#[derive(Default, Clone, serde::Deserialize, serde::Serialize)]
pub struct DoubleExpFitter {
    pub x: Vec<f64>,
    pub y: Vec<f64>,
    pub weights: Vec<f64>,
    pub fit_line_points: Vec<(f64, f64)>,
    pub upper_uncertainity_points: Vec<(f64, f64)>,
    pub lower_uncertainity_points: Vec<(f64, f64)>,
}

impl DoubleExpFitter {
    pub fn new(x: Vec<f64>, y: Vec<f64>, weights: Vec<f64>) -> Self {
        Self {
            x,
            y,
            weights,
            fit_line_points: Vec::new(),
            upper_uncertainity_points: Vec::new(),
            lower_uncertainity_points: Vec::new(),
        }
    }

    fn exponential(x: &DVector<f64>, b: f64) -> DVector<f64> {
        x.map(|x_val| (-x_val / b).exp())
    }

    fn exponential_pd_b(x: &DVector<f64>, b: f64) -> DVector<f64> {
        x.map(|x_val| (x_val / b.powi(2)) * (-x_val / b).exp())
    }

    fn exponential_pd_d(x: &DVector<f64>, d: f64) -> DVector<f64> {
        x.map(|x_val| (x_val / d.powi(2)) * (-x_val / d).exp())
    }

    pub fn double_exp_fit(&mut self, initial_b_guess: f64, initial_d_guess: f64) {
        self.fit_params = None;

        let x_data = DVector::from_vec(self.x.clone());
        let y_data = DVector::from_vec(self.y.clone());
        let weights = DVector::from_vec(self.weights.clone());

        let observation_length = x_data.len();

        let parameter_names: Vec<String> = vec!["b".to_string(), "d".to_string()];

        let initial_parameters = vec![initial_b_guess, initial_d_guess];

        let builder_proxy = SeparableModelBuilder::<f64>::new(parameter_names)
            .initial_parameters(initial_parameters)
            .independent_variable(x_data)
            .function(&["b"], Self::exponential)
            .partial_deriv("b", Self::exponential_pd_b)
            .function(&["d"], Self::exponential)
            .partial_deriv("d", Self::exponential_pd_d);

        let model = match builder_proxy.build() {
            Ok(model) => model,
            Err(err) => {
                log::error!("Error building model: {}", err);
                return;
            }
        };

        let problem = match LevMarProblemBuilder::new(model)
            .observations(y_data)
            .weights(weights)
            .build()
        {
            Ok(problem) => problem,
            Err(err) => {
                log::error!("Error building problem: {}", err);
                return;
            }
        };

        if let Ok((fit_result, fit_statistics)) =
            LevMarSolver::default().fit_with_statistics(problem)
        {
            let cov = fit_statistics.covariance_matrix();

            // Non linear parameters
            let nonlinear_parameters = fit_result.nonlinear_parameters();
            let n_nonlinear_parameters = nonlinear_parameters.len();
            let nonlinear_variances = fit_statistics.nonlinear_parameters_variance();

            // linear parameters
            let linear_coefficients = fit_result.linear_coefficients();
            let linear_coefficients = match linear_coefficients {
                Some(coefficients) => coefficients,
                None => {
                    log::error!("No linear coefficients found");
                    return;
                }
            };
            let linear_variances = fit_statistics.linear_coefficients_variance();
            let n_linear_parameters = linear_coefficients.len();

            let n_paramaters = n_nonlinear_parameters + n_linear_parameters; // total number of parameters

            // Need to get the reduced chi squared value i.e. reduced chi^2 = chi^2 / dof
            let weighted_residuals = fit_statistics.weighted_residuals();

            // Square the weighted residuals
            let squared_weighted_residuals: Vec<f64> = weighted_residuals
                .iter()
                .map(|&residual| residual * residual)
                .collect();

            // Sum up the squared weighted residuals to get the chi-squared value
            let chi_squared: f64 = squared_weighted_residuals.iter().sum();
            let dof = observation_length as f64 - n_paramaters as f64; // degrees of freedom -> N_observables - N_parameters
            let rchi2 = chi_squared / dof;

            let sigma = 1.0;
            let prob = statrs::function::erf::erf(sigma / SQRT_2); // 68.3 % confidence interval
            let scale = match statrs::distribution::StudentsT::new(0.0, 1.0, dof) {
                Ok(dist) => dist.inverse_cdf((1.0 + prob) / 2.0),
                Err(e) => {
                    log::error!("Error creating StudentsT distribution: {:?}", e);
                    return;
                }
            };

            let num_points = 1000;

            let start = 0.0;
            let end = 7000.0;

            let step = (end - start) / num_points as f64;

            let fit_points: Vec<(f64, f64)> = (0..=num_points)
                .map(|i| {
                    let x = start + i as f64 * step;
                    let y = parameter_a * (-x / parameter_b).exp()
                        + parameter_c * (-x / parameter_d).exp();

                    (x, y)
                })
                .collect();

            let confidence_band: Vec<(f64, f64)> = (0..=num_points)
                .map(|i| {
                    // followed lmfits implementation
                    let x = start + i as f64 * step;

                    let parameter_a = linear_coefficients[0];
                    let parameter_b = nonlinear_parameters[0];
                    let parameter_c = linear_coefficients[1];
                    let parameter_d = nonlinear_parameters[1];

                    let dfda = (-x / (parameter_b)).exp(); // df/da: partial derivative of a wrt parameter a
                    let dfdb = (-x / (parameter_d)).exp();  // df/db: partial derivative of a wrt parameter b
                    let dfdc = parameter_a * (x / parameter_b.powi(2)) * (-x / parameter_b).exp(); // df/dc: partial derivative of a wrt parameter c
                    let dfdd = parameter_c * (x / parameter_d.powi(2)) * (-x / parameter_d).exp();  // df/dd: partial derivative of a wrt parameter d

                    // covariance matrix has the linear parameters first then the nonlinear parameters... Hard coding it for now but it could be put in a loop
                    let y = t_dist
                        * (rchi2
                            * (dfda * dfda * cov[0]
                                + dfda * dfdb * cov[1]
                                + dfda * dfdc * cov[2]
                                + dfda * dfdd * cov[3]
                                + dfdb * dfda * cov[4]
                                + dfdb * dfdb * cov[5]
                                + dfdb * dfdc * cov[6]
                                + dfdb * dfdd * cov[7]
                                + dfdc * dfda * cov[8]
                                + dfdc * dfdb * cov[9]
                                + dfdc * dfdc * cov[10]
                                + dfdc * dfdd * cov[11]
                                + dfdd * dfda * cov[12]
                                + dfdd * dfdb * cov[13]
                                + dfdd * dfdc * cov[14]
                                + dfdd * dfdd * cov[15]))
                            .sqrt();

                    (x, y)
                })
                .collect();

            let lower_points: Vec<(f64, f64)> = fit_points
                .iter()
                .zip(confidence_band.iter())
                .map(|(fit_point, confidence_point)| {
                    (fit_point.0, fit_point.1 - confidence_point.1)
                })
                .collect();

            let upper_points: Vec<(f64, f64)> = fit_points
                .iter()
                .zip(confidence_band.iter())
                .map(|(fit_point, confidence_point)| {
                    (fit_point.0, fit_point.1 + confidence_point.1)
                })
                .collect();

            self.fit_line_points = fit_points;
            self.upper_uncertainity_points = upper_points;
            self.lower_uncertainity_points = lower_points;
        }
    }
}

```

I hope you understand what I am trying to explain! Here is two photos getting a prediction band using one and two sigma.

![image](https://github.com/geo-ant/varpro/assets/106334067/73d03fa8-baf0-4992-b69f-dd43866a5632)
![image](https://github.com/geo-ant/varpro/assets/106334067/8041e0bf-be9d-4b86-bdf8-73076c6b870c)

**Comment by geo-ant** [comment]
First of all, thanks for taking the time for the in depth explanation and example. I'm definitely interested in providing this functionality. I did make a number of smaller fixes to get the code to compile, but an expression for `t_shift` is missing. But I get the idea, since it's also layed out [here](https://www.astro.rug.nl/software/kapteyn/kmpfittutorial.html#confidence-and-prediction-intervals), right?

If I understand correctly (and forgive me, if I don't... I've been feeling a bit under the wheather lately) I'm just trying to implement this expression:

$$CB = \hat{y} \pm t_{\alpha/2, \nu}\, \sqrt{\chi_{\nu}^2\, \sum\limits_{j=0}^{j=n}\sum\limits_{k=0}^{k=n}\frac{\partial f}{\partial p_j}\frac{\partial f}{\partial p_k} C_{jk} }$$

This _should_ ™ be straightforward, right? The one quantity I don't readily have is $t_{\alpha/2, \nu}$, but I can take your lead and use the statistics crate you used. In terms of the derivatives, I should also be okay unless I am missing something. Since the user already supplied the partial derivatives to the model base functions, I can definitely use them to calculate the expressions for $\frac{\partial f}{\partial p_k}$. Either as the linear combination of the partial derivatives of the model functions if $p_k$ is a nonlinear parameter, or (if $p_k$ is a linear parameter) simply as the value of the respective model function. I don't think I would need the user to re-supply the derivatives. I've looked at how the `FitStatistics` is generated in my code and it definitely has access to the model (at least during construction). So it should be doable. 

I might have time to work on this end of this month, but I can't promise anything. I'm of course happy if you want to file a PR. If you don't want to file one that is of course fine, I might ask you for a code review if that would be okay :)

EDIT: I'm wondering how to implement this for Multiple Right Hand Sides, since varpro also supports that. In principle it should just work, I might just have to figure out a good API. But I also might run into trouble.

**Comment by geo-ant** [comment]
Hi @alconley. I had less time than I'd like so I'm not done at all. I do know how to calculate the confidence band width. But when trying to follow the calculations I started wondering whether the expression holds true for weighted least squares, too. I don't have the original source (Wolberg). The python code does not make explicit reference to weights either. I definitely used the weights for calculating the covariance matrix so they might just not explicitly appear in the formula. Or the formula might only be for unweighted least squares. Do you know more? Does lmfit support weighted least squares and would you use the confidence bands there?

**Comment by alconley** [comment]
Hi @geo-ant! Yes I have been a lot busier than expected the past couple of weeks too!  It holds true regardless of weighted or unweighted least squares fit just because we are using the covariance matrix i.e. the weights influence the covariance matrix. In most cases when I have used lmfit, it has been with weighted least squares too. 

However there is something that I do not understand yet. So fitting with and without weights on lmfit and varpro I get essentially the same fit/covariance matrix/chi-squared/reduced chi squared. However my calculation of the prediction band is not the same as lmfit when fitting _without_ weights. Lmfit predictions band fits looks reasonable whereas the one I calculate with the code above results in a prediction band with very little uncertainty. Not sure why this is happening yet.

**Comment by geo-ant** [comment]
Hi @alconley. Thanks for spending so much time on this. Are you saying that _with weitghs_ the calculations match and without weights they do not? That would be very peculiar since fitting without weights just means all weights are 1. 

The one thing I noticed glancing over your code example above is that the derivatives can't be right. df/da and df/db look like they should be the derivatives with respect to the linear parameters. However you assign the first linear parameter to a and the first nonlinear parameter to b in the lines above. That seems like it might be a bug but I have no idea whether it could be responsible for your observations.

**Comment by alconley** [comment]
All good, I plan on using this crate more in the future anyways so I am happy to help! 

So I think I figured out the problem on how to exactly match the confidence band with lmfit after some of hours of work lol. Not sure how I feel about it... but I was reading this [website](https://www.astro.rug.nl/software/kapteyn/kmpfittutorial.html#confidence-and-prediction-intervals) again but this time the reduced chi squared section. They mention scaling the uncertainty until the reduced chi squared is 1. 

Now I use the reduced chi square in calculating the uncertainty band, but instead of calculating the confidence band with the actual fit reduced chi squared, I use a value of 1. The result matches exactly for a single and double exponential fit with and without fitting with weights. This also makes sense why my double exp fit was essentially the same (a little different but close) before because the reduced chi squared of the fit was close to 1.

Hope this helps!

**Comment by geo-ant** [comment]
Hey again @alconley. Once more, this has taken me longer than expected but I have a draft working. The actual confidence bands weren't that much of an issue (that is, hoping that I did implement them correctly), but I had to rework the library API quite a bit. I think things were getting confusing for people trying to fit problems with single right hand sides (probably the normal use case), because all the things that should be vectors now became matrices. So I changed the library API to differentiate at compile time between problems with single and multiple right hand sides, so that the people working with single right hand sides get vectors back again.

I now have  a draft version ready in the feature branch `feature/confidence-bands`. If you look in the FitStatistics struct you'll find a method `confidence_band_radius` that is already documented and explains how to use it to generate confidence bands.

If (and only if) you want to, you can check it out and see if it reproduces your calculations and serves your needs. There is now also an API for calculating the best fit, so you don't have to do it manually anymore.

The functionality is not yet tested. What I want to do is what you did as well: fit the same model to the same data with lmfit and with varpro and compare the results. If you have a code snippet in Python lying around that I can use for that I'll gladly take it, if you are fine with me publishing it as part of this package. If not, that is of course completely fine and I'll write my own.

One more thing: I still have not completely grasped the explanation about the scaling and chi2 = 1. I'll have to think about that some more. Coincidentally, the original source that is quoted (Wolberg) was available for like 17 bucks via Springer, so I snatched it. Hasn't arrived yet, though...

**Comment by alconley** [comment]
Yes I would like to check it out sometime soon. However, it might be a while before I can find some time. My house got hit with a tornado! So I have a lot of stuff I have to do related to that.

Let me know if you understand the scaling and the chi2 more from the book as it also confuses me.

**Comment by geo-ant** [comment]
Hey @alconley, I'm sorry to hear that. I hope you're doing alright. Please take all the time you need, don't feel obligated to respond to anything. I'll keep you updated just in case.

**Comment by geo-ant** [comment]
Status update: I have everything working and successfully tested against the output of lmfit (both for fitting with weights and unweighted). I still need to update the docs and the readme. This might still take more time then I'd like but I'm pretty confident I'll release the new version soon (for a suitable definition of soon).

**Comment by geo-ant** [comment]
Implemented by https://github.com/geo-ant/varpro/pull/33

Feature will be released as part of the 0.9.0 release.
